satellite/repair/repairer/repairer.go

```diff
@@ -38,65 +40,106 @@ type Config struct {
 func NewService(log *zap.Logger, queue queue.RepairQueue, config *Config, repairer *SegmentRepairer) *Service {
 	return &Service{
-		log:      log,
-		queue:    queue,
-		config:   config,
-		Limiter:  sync2.NewLimiter(config.MaxRepair),
-		Loop:     *sync2.NewCycle(config.Interval),
-		repairer: repairer,
+		log:        log,
+		queue:      queue,
+		config:     config,
+		JobLimiter: semaphore.NewWeighted(int64(config.MaxRepair)),
+		Loop:       *sync2.NewCycle(config.Interval),
+		repairer:   repairer,
 	}
 }
 
 // Close closes resources
 func (service *Service) Close() error { return nil }
 
+// WaitForPendingRepairs waits for all ongoing repairs to complete.
+//
+// NB: this assumes that service.config.MaxRepair will never be changed once this Service instance
+// is initialized. If that is not a valid assumption, we should keep a copy of its initial value to
+// use here instead.
+func (service *Service) WaitForPendingRepairs() {
+	// No error return is possible here; context.Background() can't be canceled
+	_ = service.JobLimiter.Acquire(context.Background(), int64(service.config.MaxRepair))
+	service.JobLimiter.Release(int64(service.config.MaxRepair))
+}
+
 // Run runs the repairer service
 func (service *Service) Run(ctx context.Context) (err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	// wait for all repairs to complete
-	defer service.Limiter.Wait()
+	// Wait for all repairs to complete
+	defer service.WaitForPendingRepairs()
 
-	return service.Loop.Run(ctx, func(ctx context.Context) error {
-		err := service.process(ctx)
-		if err != nil {
-			service.log.Error("process", zap.Error(Error.Wrap(err)))
-		}
-		return nil
-	})
+	return service.Loop.Run(ctx, service.processWhileQueueHasItems)
 }
 
-// process picks items from repair queue and spawns a repair worker
-func (service *Service) process(ctx context.Context) (err error) {
-	defer mon.Task()(&ctx)(&err)
+// processWhileQueueHasItems keeps calling process() until the queue is empty or something
+// else goes wrong in fetching from the queue.
+func (service *Service) processWhileQueueHasItems(ctx context.Context) error {
 	for {
-		seg, err := service.queue.Select(ctx)
+		err := service.process(ctx)
 		if err != nil {
 			if storage.ErrEmptyQueue.Has(err) {
 				return nil
 			}
+			service.log.Error("process", zap.Error(Error.Wrap(err)))
 			return err
 		}
-		service.log.Info("Retrieved segment from repair queue", zap.Binary("Segment", seg.GetPath()))
+	}
+}
 
-		service.Limiter.Go(ctx, func() {
-			err := service.worker(ctx, seg)
-			if err != nil {
-				service.log.Error("repair worker failed:", zap.Binary("Segment", seg.GetPath()), zap.Error(err))
-			}
-		})
+// process picks items from repair queue and spawns a repair worker
+func (service *Service) process(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// wait until we are allowed to spawn a new job
+	if err := service.JobLimiter.Acquire(ctx, 1); err != nil {
+		return err
```

Comment:
If ab `err` gets returned from here, would it cause a deadlock on line 82?

Reply:
I don't think so. If there are already some workers going, line 82 should wait for them to finish normally and then exit. And if there aren't any going, line 82 should have no effect.

But also, I think the only possible error here comes from the context being canceled, which should mean the contexts for all repair workers should be canceled as well, so they should exit pretty soon afterward if they haven't already.